### PR TITLE
Remove user from project before deletion

### LINKED ISSUE #9974: Projects listing breaks after deleting a user added to a project.
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
UI,API
~~~

##### CLOUDSTACK VERSION

~~~
4.19.1.2
~~~

##### CONFIGURATION

Projects 

##### SUMMARY

Once you deleted a user in 'ROOT admin' role account, the projects can't be listed using UI/API. The deleted user was added to one of the projects. 

##### STEPS TO REPRODUCE

~~~
1. Add a user to the 'ROOT admin' Account.
2. Create a project as another account/user in the same domain.
3. Add the user from step 1 to this project as Project Admin.
4. As an 'admin' user delete the above user. 
~~~

##### EXPECTED RESULTS

~~~
Deleting a user shouldn't break project access. 
~~~

##### ACTUAL RESULTS

~~~
Deleting a user breaks projects access.
~~~
Undo the user deletion. ( May not be safe for production) 
Workaround example: `delete from project_account where user_id=<user_id>` OR ;`update user set state='disabled',removed=NULL where id=<id>;`

### PR DESCRIPTION
**Pull request by winterhazel**
### Description

If a user that is associated to a project gets deleted, other users become unable to see who belongs to the project via `listProjectAccounts` due to a NPE. This happens because the removed user is still associated with the project.

This PR fixes the issue by removing the user from all projects before deleting it (as we already do with accounts). To normalize environments that are affected by the bug, the following script will be executed during the upgrade in order to delete project-user associations for users that were removed.

```sql
DELETE FROM `cloud`.`project_account` WHERE `user_id` IN (SELECT `id` FROM `cloud`.`user` WHERE `removed`);
```

Fixes #9974

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [X] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

#### User clean-up on deletion

1. I created a project
2. I added a user to the project
3. I deleted the user

After step 3, I would become unable to list accounts/users that belong to the project via `listProjectAccounts` before the changes. With the changes, I was able to list them normally.

#### Normalization of affected environments

Before applying the patch, I created some broken projects by deleting their users. Then, I executed the upgrade script and verified that I was able to list their accounts/users normally.

### CODE REVIEW

**Review by DaanHoogland — APPROVED** [review]
clgtm

File: engine/schema/src/main/java/com/cloud/projects/dao/ProjectAccountDaoImpl.java [hunk 1]
```diff
@@ -194,6 +194,15 @@ public void removeAccountFromProjects(long accountId) {
         }
     }
 
+    @Override
+    public void removeUserFromProjects(long userId) {
+        SearchCriteria<ProjectAccountVO> sc = AllFieldsSearch.create();
+        sc.setParameters("userId", userId);
+
+        int count = remove(sc);
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/projects/dao/ProjectAccountDaoImpl.java:202]:
```suggestion
        int removedCount = remove(sc);
```

File: engine/schema/src/main/java/com/cloud/projects/dao/ProjectAccountDaoImpl.java [hunk 2]
```diff
@@ -194,6 +194,15 @@ public void removeAccountFromProjects(long accountId) {
...
+        s_logger.debug(String.format("Removed user [%s] from %s project(s).", userId, count));
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/projects/dao/ProjectAccountDaoImpl.java:203]:
```suggestion
        if (removedCount > 0) {
                    s_logger.debug(String.format("Removed user [%s] from %s project(s).", userId, count));
        }
```

**Review by sureshanaparti — APPROVED** [review]
minor comments, overall clgtm

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Review by rajujith — APPROVED** [review]
LGTM.
Tested, this PR fixed the reported issue.

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Pull request merged**